binding/go/porcupine.go

```diff
@@ -0,0 +1,281 @@
+// located in the "LICENSE" file accompanying this source.
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the
+// License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+// Go binding for Porcupine wake word engine. It detects utterances of given keywords within an incoming stream of
+// audio in real-time. It processes incoming audio in consecutive frames and for each frame emits the detection result.
+// The number of samples per frame can be attained by calling `.FrameLength`. The incoming audio needs to have a
+// sample rate equal to `.SampleRate` and be 16-bit linearly-encoded. Porcupine operates on single-channel audio.
+
+package porcupine
+
+import (
+	"C"
+	"embed"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+)
+
+//go:embed embedded
+var embeddedFS embed.FS
+
+// PvStatus type
+type PvStatus int
+
+// Possible status return codes from the Porcupine library
+const (
+	SUCCESS          PvStatus = 0
+	OUT_OF_MEMORY    PvStatus = 1
+	IO_ERROR         PvStatus = 2
+	INVALID_ARGUMENT PvStatus = 3
+)
+
+func pvStatusToString(status PvStatus) string {
+	switch status {
+	case SUCCESS:
+		return "SUCCESS"
+	case OUT_OF_MEMORY:
+		return "OUT_OF_MEMORY"
+	case IO_ERROR:
+		return "IO_ERROR"
+	case INVALID_ARGUMENT:
+		return "INVALID_ARGUMENT"
+	default:
+		return "Unknown error"
+	}
+}
+
+// BuiltInKeyword Type
+type BuiltInKeyword string
+
+// Available built-in wake words constants
+const (
+	ALEXA       BuiltInKeyword = "alexa"
+	AMERICANO   BuiltInKeyword = "americano"
+	BLUEBERRY   BuiltInKeyword = "blueberry"
+	BUMBLEBEE   BuiltInKeyword = "bumblebee"
+	COMPUTER    BuiltInKeyword = "computer"
+	GRAPEFRUIT  BuiltInKeyword = "grapefruit"
+	GRASSHOPPER BuiltInKeyword = "grasshopper"
+	HEY_BARISTA BuiltInKeyword = "hey barista"
+	HEY_GOOGLE  BuiltInKeyword = "hey google"
+	HEY_SIRI    BuiltInKeyword = "hey siri"
+	JARVIS      BuiltInKeyword = "jarvis"
+	OK_GOOGLE   BuiltInKeyword = "ok google"
+	PICO_CLOCK  BuiltInKeyword = "pico clock"
+	PICOVOICE   BuiltInKeyword = "picovoice"
+	PORCUPINE   BuiltInKeyword = "porcupine"
+	TERMINATOR  BuiltInKeyword = "terminator"
+)
+
+// List of available built-in wake words
+var BuiltInKeywords = []BuiltInKeyword{
+	ALEXA, AMERICANO, BLUEBERRY, BUMBLEBEE, COMPUTER, GRAPEFRUIT, GRASSHOPPER, HEY_BARISTA,
+	HEY_GOOGLE, HEY_SIRI, JARVIS, OK_GOOGLE, PICO_CLOCK, PICOVOICE, PORCUPINE, TERMINATOR,
+}
+
+// Checks if a given BuiltInKeyword is valid
+func (k BuiltInKeyword) IsValid() error {
+	for _, b := range BuiltInKeywords {
+		if k == b {
+			return nil
+		}
+	}
+	return errors.New("Invalid built-in keyword.")
+}
+
+// Porcupine struct
+type Porcupine struct {
+	handle uintptr
```

Comment:
other ones have comments so this one should probably have one too, even if its kind of self-explanatory